src/stim/search/sat/wcnf.cc

```diff
@@ -0,0 +1,261 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include "stim/search/sat/wcnf.h"
+
+
+using namespace stim;
+
+typedef double Weight;
+constexpr Weight HARD_CLAUSE_WEIGHT = -1.0;
+
+const std::string UNSAT_WCNF_STR = "p wcnf 1 2 3\n3 -1 0\n3 1 0\n";
+
+constexpr size_t BOOL_LITERAL_FALSE = SIZE_MAX - 1;
+constexpr size_t BOOL_LITERAL_TRUE = SIZE_MAX;
+
+struct BoolRef {
+  size_t variable = BOOL_LITERAL_FALSE;
+  bool negated = false;
+  BoolRef operator~() const {
+    return {variable, !negated};
+  }
+  static BoolRef False() {
+    return {BOOL_LITERAL_FALSE, false};
+  }
+  static BoolRef True() {
+    return {BOOL_LITERAL_TRUE, false};
+  }
+};
+
+struct Clause {
+  std::vector<BoolRef> vars;
+  Weight weight = HARD_CLAUSE_WEIGHT;
+  void add_var(BoolRef x) {
+    vars.push_back(x);
+  }
+};
+
+struct MaxSATInstance {
+  size_t num_variables = 0;
+  Weight max_weight = 0;
+  std::vector<Clause> clauses;
+  BoolRef new_bool() {
+    return {num_variables++};
+  }
+  void add_clause(Clause& clause) {
+    if (clause.weight != HARD_CLAUSE_WEIGHT) {
+      if (clause.weight <= 0) {
+        throw std::invalid_argument("Clauses must have positive weight or HARD_CLAUSE_WEIGHT.");
+      }
+      max_weight = std::max(max_weight, clause.weight);
+    }
+    clauses.push_back(clause);
+  }
+  BoolRef Xor(BoolRef& x, BoolRef& y) {
+    if (x.variable == BOOL_LITERAL_FALSE) {
+      return y;
+    }
+    if (x.variable == BOOL_LITERAL_TRUE) {
+      return ~y;
+    }
+    if (y.variable == BOOL_LITERAL_FALSE) {
+      return x;
+    }
+    if (y.variable == BOOL_LITERAL_TRUE) {
+      return ~x;
+    }
+    BoolRef z = new_bool();
+    // Forbid strings (x, y, z) such that z != XOR(x, y)
+    {
+      Clause clause;
+      // hard clause (x, y, z) != (0, 0, 1)
+      clause.add_var(x);
+      clause.add_var(y);
+      clause.add_var(~z);
+      add_clause(clause);
+    }
+    {
+      Clause clause;
+      // hard clause (x, y, z) != (0, 1, 0)
+      clause.add_var(x);
+      clause.add_var(~y);
+      clause.add_var(z);
+      add_clause(clause);
+    }
+    {
+      Clause clause;
+      // hard clause (x, y, z) != (1, 0, 0)
+      clause.add_var(~x);
+      clause.add_var(y);
+      clause.add_var(z);
+      add_clause(clause);
+    }
+    {
+      Clause clause;
+      // hard clause (x, y, z) != (1, 1, 1)
+      clause.add_var(~x);
+      clause.add_var(~y);
+      clause.add_var(~z);
+      add_clause(clause);
+    }
+    return z;
+  }
+
+  size_t quantized_weight(bool weighted, size_t weight_scale_factor, size_t top, Weight weight) {
+    if (weight == HARD_CLAUSE_WEIGHT) {
+      // Hard clause
+      return top;
+    }
+    // Soft clause
+    if (!weighted) {
+      // For unweighted problems, all soft clauses have weight 1.
+      return 1;
+    }
+    return std::round(weight / max_weight * (double)weight_scale_factor);
```

Comment:
This should be rounding such that it's never 0, right? It looks like this might end up as 0.

Add a unit test with an error of probability 49% as well as an error of probability 0.00000000001%.

Reply:
I added these unit tests for the shortest error problem but for the likeliest error problem it's really supposed to round down to 0 in this case.

Reply:
Won't that result in instances that appear to be unsatisfiable, despite having a solution that's merely really-unlikely?